lib/api/api_auth_webauthn.go

```diff
@@ -0,0 +1,476 @@
+	miscDBKey                      string
+	engine                         *webauthnLib.WebAuthn
+	evLogger                       events.Logger
+	userHandle                     []byte
+	registrationState              webauthnLib.SessionData
+	authenticationState            webauthnLib.SessionData
+	credentialsPendingRegistration []WebauthnCredential
+}
+
+func newWebauthnService(guiCfg config.GUIConfiguration, deviceName string, evLogger events.Logger, miscDB *db.NamespacedKV, miscDBKey string) (webauthnService, error) {
+	engine, err := newWebauthnEngine(guiCfg, deviceName)
+	if err != nil {
+		return webauthnService{}, err
+	}
+
+	userHandle, err := base64.URLEncoding.DecodeString(guiCfg.WebauthnUserId)
+	if err != nil {
+		return webauthnService{}, err
+	}
+
+	return webauthnService{
+		miscDB:     miscDB,
+		miscDBKey:  miscDBKey,
+		engine:     engine,
+		evLogger:   evLogger,
+		userHandle: userHandle,
+	}, nil
+}
+
+func (s *webauthnService) loadState() (WebauthnState, error) {
+	stateBytes, ok, err := s.miscDB.Bytes(s.miscDBKey)
+	if err != nil {
+		return WebauthnState{}, err
+	}
+	if !ok {
+		return WebauthnState{}, nil
+	}
+
+	var state WebauthnState
+	err = state.Unmarshal(stateBytes)
+	if err != nil {
+		return WebauthnState{}, err
+	}
+
+	return state, nil
+}
+
+func (s *webauthnService) storeState(state WebauthnState) error {
+	stateBytes, err := state.Marshal()
+	if err != nil {
+		return err
+	}
+
+	return s.miscDB.PutBytes(s.miscDBKey, stateBytes)
+}
+
+func (s *WebauthnState) Copy() WebauthnState {
+	c := *s
+	c.Credentials = make([]WebauthnCredential, len(s.Credentials))
+	for i := range s.Credentials {
+		c.Credentials[i] = s.Credentials[i].Copy()
+	}
+	return c
+}
+
+func (g *WebauthnCredential) Copy() WebauthnCredential {
+	c := *g
+	if c.Transports != nil {
+		c.Transports = make([]string, len(c.Transports))
+		copy(c.Transports, g.Transports)
+	}
+	return c
+}
+
+func (c *WebauthnCredential) NicknameOrID() string {
+	if c.Nickname != "" {
+		return c.Nickname
+	} else {
+		return c.ID
+	}
+}
+
+func (s *webauthnService) user(guiCfg config.GUIConfiguration) webauthnLibUser {
+	return webauthnLibUser{
+		service: s,
+		guiCfg:  guiCfg,
+	}
+}
+
+type webauthnLibUser struct {
+	service *webauthnService
+	guiCfg  config.GUIConfiguration
+}
+
+func (u webauthnLibUser) WebAuthnID() []byte {
+	return u.service.userHandle
+}
+func (u webauthnLibUser) WebAuthnName() string {
+	return u.guiCfg.User
+}
+func (u webauthnLibUser) WebAuthnDisplayName() string {
+	return u.guiCfg.User
+}
+func (webauthnLibUser) WebAuthnIcon() string {
+	return ""
+}
+func (u webauthnLibUser) WebAuthnCredentials() []webauthnLib.Credential {
+	var result []webauthnLib.Credential
+	eligibleCredentials, err := u.service.EligibleWebAuthnCredentials(u.guiCfg)
+	if err != nil {
+		return make([]webauthnLib.Credential, 0)
+	}
+
+	for _, cred := range eligibleCredentials {
+		id, err := base64.URLEncoding.DecodeString(cred.ID)
```

Comment:
The main reason for this is that these were originally stored in `config.xml`, and https://github.com/emlun/syncthing/pull/7 (motivated by https://github.com/syncthing/syncthing/pull/9175#issuecomment-2263931032, see: https://github.com/syncthing/syncthing/pull/9175#issuecomment-2282943185) would move them back there. The XML encoder/decoder does not seem to text-encode binary values, rather it just dumps the raw binary (with XML control chars escaped) into the XML stream, so that would make `config.xml` awkward to work with in a text editor.

https://github.com/emlun/syncthing/pull/6 is related - that's also a binary value, but we (the server) get to choose that value (unlike the credential ID, public key etc.) quite freely, so we can choose a value that consists only of, say, printable ASCII (in this case, only `[0-9a-f]`) even when dumped as binary.

Reply:
It might be worth it to separate the config structs and the API structs. Then the config can be handled specially and the rest comes automatically.

Reply:
Is this still relevant now with https://github.com/emlun/syncthing/pull/7 merged? In the case pointed out here, these values are coming directly from config and thus we don't get away from having to be decode here before we can pass them into the `go-webauthn` library.